File: solutions/go/bob/1/bob.go
// Package bob provides a simulation of a lackadaisical teenager's responses.
package bob

import "strings"

// Hey returns Bob's response to various types of remarks.
func Hey(remark string) string {
	remark = strings.TrimSpace(remark)

	if remark == "" {
		return "Fine. Be that way!"
	}

	isQuestion := strings.HasSuffix(remark, "?")
	isShouting := hasLetters(remark) && strings.ToUpper(remark) == remark

	if isShouting && isQuestion {
		return "Calm down, I know what I'm doing!"
	}

	if isShouting {
		return "Whoa, chill out!"
	}

	if isQuestion {
		return "Sure."
	}

	return "Whatever."
}

func hasLetters(s string) bool {
	return strings.ContainsAny(strings.ToLower(s), "abcdefghijklmnopqrstuvwxyz")
}

File: solutions/cpp/bob/1/bob.cpp
#include "bob.h"
#include <algorithm>
#include <cctype>

namespace bob {

std::string hey(const std::string& utterance) {
    size_t first = utterance.find_first_not_of(" \t\n\r");
    if (first == std::string::npos) {
        return "Fine. Be that way!";
    }
    
    size_t last = utterance.find_last_not_of(" \t\n\r");
    std::string cleaned = utterance.substr(first, last - first + 1);

    bool is_question = (cleaned.back() == '?');
    bool has_letters = std::any_of(cleaned.begin(), cleaned.end(), [](unsigned char c) { 
        return std::isalpha(c); 
    });
    bool is_yelling = has_letters && std::none_of(cleaned.begin(), cleaned.end(), [](unsigned char c) { 
        return std::islower(c); 
    });

    if (is_yelling && is_question) {
        return "Calm down, I know what I'm doing!";
    }
    if (is_yelling) {
        return "Whoa, chill out!";
    }
    if (is_question) {
        return "Sure.";
    }

    return "Whatever.";
}

}  // namespace bob
